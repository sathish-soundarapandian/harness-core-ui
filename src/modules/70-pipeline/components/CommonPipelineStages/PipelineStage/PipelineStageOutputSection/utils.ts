/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { AllowedTypes, getMultiTypeFromValue, MultiTypeInputType } from '@harness/uicore'
import type { FormikErrors, FormikProps } from 'formik'
import { get, isEmpty, set } from 'lodash-es'
import type { UseStringsReturn } from 'framework/strings'
import type { PipelineStageOutputs } from 'services/pipeline-ng'

export const MAX_LENGTH = 64

export interface PipelineStageOutputSectionProps {
  children: React.ReactElement
}

export interface PipelineStageOutputData {
  outputs: PipelineStageOutputs[]
}

export interface OutputPanelInputSetProps {
  template: PipelineStageOutputData
  path: string
  allowableTypes: AllowedTypes
  readonly?: boolean
}

export interface ConectedOutputPanelInputSetProps extends OutputPanelInputSetProps {
  formik: FormikProps<PipelineStageOutputData>
}

interface ValidateOutputPanelInputSetProp {
  data: PipelineStageOutputData
  template?: PipelineStageOutputData
  getString?: UseStringsReturn['getString']
}

export const validateOutputPanelInputSet = ({
  data,
  template,
  getString
}: ValidateOutputPanelInputSetProp): FormikErrors<PipelineStageOutputData> => {
  const errors: FormikErrors<PipelineStageOutputData> = { outputs: [] }
  data?.outputs?.forEach((output: PipelineStageOutputs, index: number) => {
    const currentVariableTemplate = get(template, `outputs[${index}].value`, '')
    if (isEmpty(output.value) && getMultiTypeFromValue(currentVariableTemplate) === MultiTypeInputType.RUNTIME) {
      set(errors, `outputs[${index}].value`, getString?.('fieldRequired', { field: output.name }))
    }
  })
  return errors
}
