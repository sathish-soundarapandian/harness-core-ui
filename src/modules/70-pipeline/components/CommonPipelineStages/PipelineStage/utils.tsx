/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

/* eslint-disable react/function-component-definition */
import React from 'react'
import type { CellProps, Renderer } from 'react-table'
import {
  Text,
  FontVariation,
  Layout,
  Icon,
  Color,
  Popover,
  ButtonSize,
  ButtonVariation,
  Button,
  IconName
} from '@harness/uicore'
import { defaultTo, get } from 'lodash-es'
import { Classes, PopoverInteractionKind, Position } from '@blueprintjs/core'
import { useParams } from 'react-router-dom'
import { String, useStrings } from 'framework/strings'
import type { PMSPipelineSummaryResponse } from 'services/pipeline-ng'
import { useAppStore } from 'framework/AppStore/AppStoreContext'
import { StoreType } from '@common/constants/GitSyncTypes'
import routes from '@common/RouteDefinitions'
import type { PipelineType, ProjectPathProps } from '@common/interfaces/RouteInterfaces'
import css from './PipelineStageMinimalMode.module.scss'

interface CodeSourceWrapper {
  textName: string
  iconName: IconName
  size: number
}

export const PipelineNameIdTagCell: Renderer<CellProps<PMSPipelineSummaryResponse>> = ({ row }) => {
  const data = row.original
  return (
    <Layout.Horizontal spacing="xsmall">
      <Layout.Vertical>
        <Text font={{ variation: FontVariation.H6 }} lineClamp={1} color={Color.BLACK}>
          {data?.name}
        </Text>
        <Text font={{ variation: FontVariation.BODY }} lineClamp={1} color={Color.GREY_600}>
          <String stringID="idLabel" vars={{ id: data?.identifier }} />
        </Text>
      </Layout.Vertical>
      {Object.keys(get(data, 'tags', {})).length > 0 && (
        <>
          <Icon name="main-tags" size={15} />
          <Text color={Color.GREY_600}>{Object.keys(get(data, 'tags', {})).length}</Text>
        </>
      )}
    </Layout.Horizontal>
  )
}

export const CodeSourceCell: Renderer<CellProps<PMSPipelineSummaryResponse>> = ({ row }) => {
  const { gitDetails } = row.original
  const { getString } = useStrings()
  const data = row.original
  const { isGitSyncEnabled: isGitSyncEnabledForProject, gitSyncEnabledOnlyForFF } = useAppStore()
  const isGitSyncEnabled = isGitSyncEnabledForProject && !gitSyncEnabledOnlyForFF
  const isRemote = data.storeType === StoreType.REMOTE || isGitSyncEnabled
  const inlineWrapper: CodeSourceWrapper = {
    textName: getString('inline'),
    iconName: 'repository',
    size: 10
  }
  const remoteWrapper: CodeSourceWrapper = {
    textName: getString('repository'),
    iconName: 'remote-setup',
    size: 12
  }

  return (
    <div className={css.codeSourceColumnContainer}>
      <Popover
        disabled={!isRemote}
        position={Position.TOP}
        interactionKind={PopoverInteractionKind.HOVER}
        className={Classes.DARK}
        content={
          <Layout.Vertical spacing="small" padding="large" className={css.contentWrapper}>
            <Layout.Horizontal spacing="small" flex={{ alignItems: 'center', justifyContent: 'start' }}>
              <Icon name="github" size={14} color={Color.GREY_200} />
              <Text color={Color.WHITE} font={{ variation: FontVariation.SMALL }}>
                {get(gitDetails, 'repoName', get(gitDetails, 'repoIdentifier'))}
              </Text>
            </Layout.Horizontal>
            {gitDetails?.filePath && (
              <Layout.Horizontal spacing="small" flex={{ alignItems: 'center', justifyContent: 'start' }}>
                <Icon name="remotefile" size={14} color={Color.GREY_200} />
                <Text color={Color.WHITE} font={{ variation: FontVariation.SMALL }}>
                  {gitDetails.filePath}
                </Text>
              </Layout.Horizontal>
            )}
          </Layout.Vertical>
        }
      >
        <div className={css.codeSourceColumn}>
          <Icon
            name={isRemote ? remoteWrapper.iconName : inlineWrapper.iconName}
            size={isRemote ? remoteWrapper.size : inlineWrapper.size}
            color={Color.GREY_600}
          />
          <Text margin={{ left: 'xsmall' }} font={{ variation: FontVariation.TINY_SEMI }} color={Color.GREY_600}>
            {isRemote ? remoteWrapper.textName : inlineWrapper.textName}
          </Text>
        </div>
      </Popover>
    </div>
  )
}

export const ViewPipelineButtonCell: Renderer<CellProps<PMSPipelineSummaryResponse>> = ({ row, column }) => {
  const { getString } = useStrings()
  const { accountId, module } = useParams<PipelineType<ProjectPathProps>>()
  const data = row.original
  const orgIdentifier = (column as any)?.orgIdentifier
  const projectIdentifier = (column as any)?.projectIdentifier

  const openPipelineInNewTab = (): void => {
    const pipelineStudioPath = routes.toPipelineStudio({
      projectIdentifier,
      orgIdentifier,
      accountId: defaultTo(accountId, ''),
      pipelineIdentifier: get(data, 'identifier', ''),
      module,
      repoIdentifier: get(data, 'gitDetails.repoIdentifier'),
      connectorRef: get(data, 'connectorRef'),
      repoName: get(data, 'gitDetails.repoName'),
      branch: get(data, 'gitDetails.branch'),
      storeType: get(data, 'storeType') as StoreType
    })
    window.open(`${window.location.origin}${window.location.pathname}#${pipelineStudioPath}`, '_blank')
  }

  return (
    <Button
      variation={ButtonVariation.SECONDARY}
      size={ButtonSize.SMALL}
      text={getString('common.viewText')}
      icon="main-view"
      onClick={e => {
        e.preventDefault()
        e.stopPropagation()
        openPipelineInNewTab()
      }}
    />
  )
}
