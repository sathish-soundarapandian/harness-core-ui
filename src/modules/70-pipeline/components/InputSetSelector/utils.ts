/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import type { IconName, SelectOption } from '@harness/uicore'
import type { EntityGitDetails, InputSetSummaryResponse } from 'services/pipeline-ng'
import css from './InputSetSelector.module.scss'

type InputSetLocal = InputSetSummaryResponse & SelectOption
export interface InputSetValue extends InputSetLocal {
  type: InputSetSummaryResponse['inputSetType']
  gitDetails?: EntityGitDetails
}
export interface ChildPipelineStageProps {
  childOrgIdentifier: string
  childProjectIdentifier: string
  usePortal?: boolean
  inputSetReferences?: string[]
}

export const getIconByType = (type: InputSetSummaryResponse['inputSetType']): IconName => {
  return type === 'OVERLAY_INPUT_SET' ? 'step-group' : 'yaml-builder-input-sets'
}

export const onDragStart = (event: React.DragEvent<HTMLLIElement>, row: InputSetValue): void => {
  event.dataTransfer.setData('data', JSON.stringify(row))
  event.currentTarget.classList.add(css.dragging)
}

export const onDragEnd = (event: React.DragEvent<HTMLLIElement>): void => {
  event.currentTarget.classList.remove(css.dragging)
}

export const onDragLeave = (event: React.DragEvent<HTMLLIElement>): void => {
  event.currentTarget.classList.remove(css.dragOver)
}

export const onDragOver = (event: React.DragEvent<HTMLLIElement>): void => {
  if (event.preventDefault) {
    event.preventDefault()
  }
  event.currentTarget.classList.add(css.dragOver)
  event.dataTransfer.dropEffect = 'move'
}
