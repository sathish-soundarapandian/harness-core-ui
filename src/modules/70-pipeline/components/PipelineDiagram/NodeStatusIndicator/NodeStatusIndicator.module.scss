/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

.stepCount {
  display: inline-flex;
  align-items: center;
  padding: 0 var(--spacing-2) !important;
  font-size: var(--font-size-normal) !important;
  color: var(--grey-900);
  font-weight: 500;
  height: var(--spacing-6);
  white-space: nowrap;

  & > :global(.bp3-icon),
  & > .running-animation {
    margin-right: var(--spacing-2);
  }

  &[data-status='success'] > :global(.bp3-icon) {
    color: var(--green-500);
  }

  &[data-status='failed'] > :global(.bp3-icon) {
    color: var(--red-500);
  }
}

.stepCountWrapper {
  .stepCount {
    border-left: 1px solid var(--grey-200);
  }
  :first-child {
    border-left: 0px !important;
  }
}
