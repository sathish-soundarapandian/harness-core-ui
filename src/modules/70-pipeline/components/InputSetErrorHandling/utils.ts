/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import type { MutateMethod } from 'restful-react'
import { defaultTo, get } from 'lodash-es'
import { useHistory, useParams } from 'react-router-dom'
import { useToaster } from '@harness/uicore'
import type { InputSetDTO, SaveInputSetDTO, StatusType } from '@pipeline/utils/types'
import { AppStoreContext } from 'framework/AppStore/AppStoreContext'
import type {
  EntityGitDetails,
  ResponseInputSetResponse,
  ResponseOverlayInputSetResponse,
  UpdateInputSetForPipelinePathParams,
  UpdateInputSetForPipelineQueryParams,
  UpdateOverlayInputSetForPipelinePathParams,
  UpdateOverlayInputSetForPipelineQueryParams
} from 'services/pipeline-ng'
import type { SaveToGitFormInterface } from '@common/components/SaveToGitForm/SaveToGitForm'
import { StoreMetadata, StoreType } from '@common/constants/GitSyncTypes'
import { UseSaveSuccessResponse, useSaveToGitDialog } from '@common/modals/SaveToGitDialog/useSaveToGitDialog'
import type { InputSetGitQueryParams, InputSetPathProps, PipelineType } from '@common/interfaces/RouteInterfaces'
import { useQueryParams } from '@common/hooks'
import type { SaveToGitFormV2Interface } from '@common/components/SaveToGitFormV2/SaveToGitFormV2'
import { useStrings } from 'framework/strings'
import { yamlStringify } from '@common/utils/YamlHelperMethods'
import { clearNullUndefined } from '@pipeline/utils/inputSetUtils'
import useRBACError from '@rbac/utils/useRBACError/useRBACError'
import type {
  OverlayInputSetDTO,
  SaveOverlayInputSetDTO
} from '@pipeline/components/OverlayInputSetForm/OverlayInputSetForm'

type UpdateInpSetOrOverlayInpSetReturnType = Promise<{
  status?: StatusType
  nextCallback: () => void
}>

interface UseSaveInputSetOrOverlayInpSetReturnType {
  handleSubmit: (inputSetObj: InputSetDTO | OverlayInputSetDTO, storeMetadata?: StoreMetadata) => Promise<void>
}

interface InputSetInfo {
  updateInputSet: MutateMethod<
    ResponseInputSetResponse,
    string,
    UpdateInputSetForPipelineQueryParams,
    UpdateInputSetForPipelinePathParams
  >
  updateOverlayInputSet: MutateMethod<
    ResponseOverlayInputSetResponse,
    void,
    UpdateOverlayInputSetForPipelineQueryParams,
    UpdateOverlayInputSetForPipelinePathParams
  >
  inputSet: InputSetDTO | OverlayInputSetDTO
  _isOverlayInputSet: boolean
  refetch?: () => Promise<void>
  fromInputSetForm?: boolean
  hideForm?: () => void
  inpSetGitDetails?: EntityGitDetails
  onReconcile?: (identifier: string) => void
  refetchInputSets?: () => void
}

export function useSaveInputSetOrOverlayInpSet(inputSetInfo: InputSetInfo): UseSaveInputSetOrOverlayInpSetReturnType {
  const {
    updateInputSet,
    updateOverlayInputSet,
    inputSet,
    _isOverlayInputSet,
    refetch,
    fromInputSetForm,
    hideForm,
    inpSetGitDetails,
    onReconcile,
    refetchInputSets
  } = inputSetInfo
  const { getString } = useStrings()
  const { showSuccess, showError } = useToaster()
  const { getRBACErrorMessage } = useRBACError()
  const history = useHistory()
  const { isGitSyncEnabled } = React.useContext(AppStoreContext)

  const { projectIdentifier, orgIdentifier, accountId, pipelineIdentifier } = useParams<
    PipelineType<InputSetPathProps> & { accountId: string }
  >()
  const { repoIdentifier, branch, repoName, connectorRef, storeType } = useQueryParams<InputSetGitQueryParams>()

  const [savedInputSetObj, setSavedInputSetObj] = React.useState<InputSetDTO | OverlayInputSetDTO>({})
  const [initialGitDetails, setInitialGitDetails] = React.useState<EntityGitDetails>({ repoIdentifier, branch })
  const [initialStoreMetadata, setInitialStoreMetadata] = React.useState<StoreMetadata>({
    repoName,
    branch,
    connectorRef,
    storeType
  })

  const isInline = !isGitSyncEnabled && initialStoreMetadata.storeType !== StoreType.REMOTE

  /* istanbul ignore next */ const handleReconcile = (): void => {
    if (initialStoreMetadata.storeType === StoreType.REMOTE) {
      onReconcile?.(get(inputSet, 'identifier', ''))
    }
    refetch?.()
    hideForm?.()
    refetchInputSets?.()
  }

  const updateInpSetOrOverlayInpSet = React.useCallback(
    async (
      inputSetObj: InputSetDTO | OverlayInputSetDTO,
      gitDetails?: SaveToGitFormInterface,
      objectId = ''
    ): UpdateInpSetOrOverlayInpSetReturnType => {
      let response: ResponseInputSetResponse | ResponseOverlayInputSetResponse | null = null
      try {
        /* istanbul ignore else */ if (inputSetObj.identifier) {
          if (_isOverlayInputSet) {
            response = await updateOverlayInputSet(
              yamlStringify({ overlayInputSet: clearNullUndefined(inputSetObj) }) as unknown as void,
              {
                pathParams: {
                  inputSetIdentifier: defaultTo(inputSetObj.identifier, '')
                },
                queryParams: {
                  accountIdentifier: accountId,
                  orgIdentifier,
                  pipelineIdentifier: pipelineIdentifier ?? get(inputSet, 'pipelineIdentifier'),
                  projectIdentifier,
                  ...(isGitSyncEnabled
                    ? {
                        pipelineRepoID: repoIdentifier,
                        pipelineBranch: branch
                      }
                    : {}),
                  ...(initialStoreMetadata.storeType === StoreType.REMOTE ? initialStoreMetadata : {}),
                  ...(gitDetails
                    ? { ...gitDetails, lastObjectId: objectId, lastCommitId: initialGitDetails.commitId }
                    : {}),
                  ...(gitDetails && gitDetails.isNewBranch ? { baseBranch: initialGitDetails.branch } : {})
                }
              }
            )
          } else {
            response = await updateInputSet(yamlStringify({ inputSet: clearNullUndefined(inputSetObj) }), {
              pathParams: {
                inputSetIdentifier: defaultTo(inputSetObj.identifier, '')
              },
              queryParams: {
                accountIdentifier: accountId,
                orgIdentifier,
                pipelineIdentifier: pipelineIdentifier ?? get(inputSet, 'pipelineIdentifier'),
                projectIdentifier,
                ...(isGitSyncEnabled
                  ? {
                      pipelineRepoID: repoIdentifier,
                      pipelineBranch: branch
                    }
                  : {}),
                ...(initialStoreMetadata.storeType === StoreType.REMOTE ? initialStoreMetadata : {}),
                ...(gitDetails
                  ? { ...gitDetails, lastObjectId: objectId, lastCommitId: initialGitDetails.commitId }
                  : {}),
                ...(gitDetails && gitDetails.isNewBranch ? { baseBranch: initialGitDetails.branch } : {})
              }
            })
          }
          /* istanbul ignore else */ if (get(response, 'data') && isInline) refetch?.()
        } else {
          throw new Error(getString('common.validation.identifierIsRequired'))
        }
        showSuccess(
          _isOverlayInputSet
            ? getString('pipeline.inputSets.overlayISUpdated', { name: get(inputSet, 'name', '') })
            : getString('pipeline.inputSets.inputSetUpdated', { name: get(inputSet, 'name', '') })
        )
        /* istanbul ignore else */ if (isInline) {
          !!fromInputSetForm && history.goBack()
          hideForm?.()
          refetchInputSets?.()
        }
      } catch (e) {
        if (isInline) {
          showError(getRBACErrorMessage(e), undefined, 'pipeline.common.error')
        } else {
          throw e
        }
      }
      return {
        status: response?.status,
        nextCallback: /* istanbul ignore next */ () => (fromInputSetForm ? history.goBack() : handleReconcile())
      }
    },
    [
      _isOverlayInputSet,
      inputSet,
      isGitSyncEnabled,
      initialStoreMetadata,
      refetch,
      updateOverlayInputSet,
      accountId,
      orgIdentifier,
      pipelineIdentifier,
      projectIdentifier,
      repoIdentifier,
      branch,
      initialGitDetails.commitId,
      initialGitDetails.branch,
      updateInputSet,
      fromInputSetForm,
      history,
      hideForm,
      refetchInputSets
    ]
  )

  const { openSaveToGitDialog } = useSaveToGitDialog<SaveInputSetDTO | SaveOverlayInputSetDTO>({
    onSuccess: (
      gitData: SaveToGitFormInterface & SaveToGitFormV2Interface,
      payload?: SaveInputSetDTO | SaveOverlayInputSetDTO,
      objectId?: string
    ): Promise<UseSaveSuccessResponse> =>
      updateInpSetOrOverlayInpSet(
        defaultTo(
          _isOverlayInputSet
            ? (payload as SaveOverlayInputSetDTO)?.overlayInputSet
            : (payload as SaveInputSetDTO)?.inputSet,
          savedInputSetObj
        ),
        gitData,
        objectId
      )
  })

  const handleSubmit = React.useCallback(
    async (inputSetObj: InputSetDTO | OverlayInputSetDTO, storeMetadata?: StoreMetadata) => {
      const _gitDetails = fromInputSetForm ? get(inputSet, 'gitDetails') : inpSetGitDetails
      setSavedInputSetObj(inputSetObj)
      setInitialGitDetails(defaultTo(_gitDetails, {}))
      setInitialStoreMetadata(defaultTo(storeMetadata, {}))

      /* istanbul ignore else */ if (inputSetObj) {
        _isOverlayInputSet && delete inputSetObj.pipeline
        if (isGitSyncEnabled || get(storeMetadata, 'storeType') === StoreType.REMOTE) {
          openSaveToGitDialog({
            isEditing: true,
            resource: {
              type: 'InputSets',
              name: inputSetObj.name as string,
              identifier: inputSetObj.identifier as string,
              gitDetails: _gitDetails,
              storeMetadata: get(storeMetadata, 'storeType') === StoreType.REMOTE ? storeMetadata : undefined
            },
            payload: _isOverlayInputSet ? { overlayInputSet: inputSetObj } : { inputSet: inputSetObj }
          })
        } else {
          updateInpSetOrOverlayInpSet(inputSetObj)
        }
      }
    },
    [
      _isOverlayInputSet,
      isGitSyncEnabled,
      inputSet,
      updateInpSetOrOverlayInpSet,
      openSaveToGitDialog,
      inpSetGitDetails,
      fromInputSetForm
    ]
  )

  return {
    handleSubmit
  }
}
