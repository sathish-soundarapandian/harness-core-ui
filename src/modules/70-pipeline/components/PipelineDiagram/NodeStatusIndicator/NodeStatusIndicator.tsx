/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { Icon, Layout } from '@harness/uicore'
import { defaultTo } from 'lodash-es'
import cx from 'classnames'
import React from 'react'
import { ExecutionStatusEnum } from '@pipeline/utils/statusHelpers'
import type { PipelineGraphState } from '../types'
import css from './NodeStatusIndicator.module.scss'

interface NodeStatusIndicatorProps {
  nodeState: PipelineGraphState[]
  className?: string
}
export function NodeStatusIndicator(props: NodeStatusIndicatorProps): JSX.Element {
  const { nodeState, className = '' } = props
  let successfulStagesCount = 0
  let failedStagesCount = 0
  let runningStagesCount = 0

  nodeState?.forEach((stateVal: PipelineGraphState & any) => {
    const nodeStatus = defaultTo(stateVal?.status, stateVal?.step?.status)

    if (nodeStatus === ExecutionStatusEnum.Success) {
      successfulStagesCount += 1
    } else if (nodeStatus === ExecutionStatusEnum.Failed) {
      failedStagesCount += 1
    } else if (nodeStatus === ExecutionStatusEnum.Running) {
      runningStagesCount += 1
    }
  })
  return (
    <Layout.Horizontal padding={0} className={cx(css.stepCountWrapper, className)}>
      <div className={css.stepCount} data-status="success">
        <Icon name={'success-tick'} size={14} />
        {successfulStagesCount}
      </div>
      {runningStagesCount > 0 && (
        <div className={css.stepCount} data-status="running">
          <Icon name={'loading'} size={14} />
          {runningStagesCount}
        </div>
      )}
      <div className={css.stepCount} data-status="failed">
        <Icon name={'warning-sign'} size={14} data-status="failed" />
        {failedStagesCount}
      </div>
    </Layout.Horizontal>
  )
}
