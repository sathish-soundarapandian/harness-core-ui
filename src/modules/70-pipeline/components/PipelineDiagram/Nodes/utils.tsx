/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { isNumber } from 'lodash-es'
import { yamlStringify } from '@common/utils/YamlHelperMethods'
import { dragPlaceholderImageBase64 } from './assets/dragImageBase64'
import { dragStagePlaceholderImageBase64 } from './assets/dragStageImageBase64'
import type { Dimension } from './NodeDimensionStore'

export enum NodeEntity {
  STAGE = 'STAGE',
  STEP = 'STEP'
}
export interface LayoutStyles extends Pick<Dimension, 'height' | 'width'> {
  marginLeft?: string
}

export const COLLAPSED_MATRIX_NODE_LENGTH = 8
export const MAX_ALLOWED_MATRIX_COLLAPSED_NODES = 4
export const DEFAULT_MATRIX_PARALLELISM = 1

export const getPositionOfAddIcon = (props: any, isRightNode?: boolean): string => {
  if (isRightNode) {
    return '-40px'
  }
  if (props?.children?.length) {
    if (props?.prevNode?.children?.length) {
      return '-65px'
    }
    if (props?.prevNode) return '-58px'
  }
  if (props?.prevNode?.children?.length) {
    return '-58px'
  }
  if (props?.parentIdentifier && !props.prevNode) {
    return '-35px'
  }
  return '-50px'
}

export const transformMatrixLabels = (nodeData: string): string => {
  const parsedNodeName = {}
  try {
    // if object parse to yaml
    const parsedVal = JSON.parse(nodeData)
    if (parsedVal && !isNumber(parsedVal)) {
      Object.assign(parsedNodeName, parsedVal)
      return yamlStringify(parsedNodeName, { indent: 4 })
    } else {
      return JSON.stringify(nodeData)
    }
  } catch (_e) {
    // name is string/number, parse it to string
    return JSON.stringify(nodeData)
  }
}

export function getSGDimensions(nodeDimensionMetaData: Dimension, index: number): LayoutStyles {
  if (nodeDimensionMetaData?.isNodeCollapsed) {
    return {
      height: 118,
      width: 134
    }
  }
  const height = nodeDimensionMetaData?.height + 68 + (index > 0 ? 60 : 0)
  const width = nodeDimensionMetaData?.width + 82 + (index > 0 ? 80 : 0)

  return { height, width }
}

export const getMatrixHeight = (
  nodeHeight: number,
  maxChildLength: number,
  parallelism: number,
  showAllNodes: boolean
): number => {
  if (parallelism === 0 || maxChildLength === parallelism) {
    // parallel case
    return maxChildLength * nodeHeight
  } else if (!showAllNodes && maxChildLength < parallelism) {
    // collapsed mode, single row
    return nodeHeight
  } else {
    return (
      (Math.floor(maxChildLength / parallelism) + Math.ceil((maxChildLength % parallelism) / parallelism)) * nodeHeight
    )
  }
}

export const attachDragImageToEventHandler = (event: React.DragEvent<HTMLDivElement>, type?: NodeEntity): void => {
  // set drag image preview to custom icon in case of safari browser, as safari blocks image preview if dom tree have transform property on any parent
  if (navigator.userAgent.search('Safari') >= 0 && navigator.userAgent.search('Chrome') < 0) {
    const dragImage = document.createElement('img')
    dragImage.src = type === NodeEntity.STAGE ? dragStagePlaceholderImageBase64 : dragPlaceholderImageBase64
    dragImage.style.position = 'absolute'
    dragImage.style.left = '-100%'

    document.body.append(dragImage)
    setTimeout(() => dragImage.remove(), 0)

    event.dataTransfer.setDragImage(dragImage, dragImage.width / 2, dragImage.height / 2)
  }
}
