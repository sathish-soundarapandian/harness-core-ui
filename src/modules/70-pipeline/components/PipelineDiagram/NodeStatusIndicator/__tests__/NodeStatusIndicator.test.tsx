/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import { render } from '@testing-library/react'
import { TestWrapper } from '@common/utils/testUtils'
import { NodeStatusIndicator } from '../NodeStatusIndicator'

describe('Node Status Indicator Test', () => {
  // eslint-disable-next-line jest/no-disabled-tests
  test('Should render diagramnodes', () => {
    const { container } = render(
      <TestWrapper>
        <NodeStatusIndicator nodeState={[{ status: 'Success' }, { status: 'Failed' }, { status: 'Running' } as any]} />
      </TestWrapper>
    )
    const successStatusLabel = container.querySelector(`[data-icon="success-tick"]`)
    const failedStatusLabel = container.querySelector(`[data-icon="failed"]`)
    const runninngStatusLabel = container.querySelector(`[data-icon="running"]`)
    expect(successStatusLabel).toBeDefined()
    expect(failedStatusLabel).toBeDefined()
    expect(runninngStatusLabel).toBeDefined()
    expect(container).toMatchSnapshot()
  })
})
