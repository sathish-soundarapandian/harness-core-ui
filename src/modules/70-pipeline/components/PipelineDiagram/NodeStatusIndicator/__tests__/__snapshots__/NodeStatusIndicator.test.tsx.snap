// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Node Status Indicator Test Should render diagramnodes 1`] = `
<div>
  <div
    class="Layout--horizontal StyledProps--main stepCountWrapper"
  >
    <div
      class="stepCount"
      data-status="success"
    >
      <span
        class="bp3-icon StyledProps--main"
        data-icon="success-tick"
      >
        <svg
          fill="none"
          height="14"
          viewBox="0 0 32 32"
          width="14"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M16 28.995c-7.18 0-13-5.82-13-13s5.82-13 13-13 13 5.82 13 13-5.82 13-13 13zm7.261-16.118a1 1 0 10-1.327-1.496l-6.82 6.049-3.242-3.206a1 1 0 10-1.406 1.422l3.909 3.864.665.659.701-.622 7.52-6.67z"
            fill="#42AB45"
            fill-rule="evenodd"
          />
          <path
            clip-rule="evenodd"
            d="M15.004 21.538l8.92-7.913a2 2 0 10-2.653-2.992l.663.748a1 1 0 111.327 1.496l-8.22 7.292-4.575-4.523a1 1 0 111.406-1.422l3.243 3.206 6.82-6.05-.664-.747-6.12 5.427-2.576-2.548a2 2 0 00-2.813 2.845l5.242 5.181zM16 27.995c-6.627 0-12-5.373-12-12 0-6.628 5.373-12 12-12 6.628 0 12 5.372 12 12 0 6.627-5.372 12-12 12zm-13-12c0 7.18 5.82 13 13 13s13-5.82 13-13-5.82-13-13-13-13 5.82-13 13z"
            fill="#42AB45"
            fill-rule="evenodd"
          />
        </svg>
      </span>
      1
    </div>
    <div
      class="stepCount"
      data-status="running"
    >
      <span
        class="bp3-icon StyledProps--main"
        data-icon="loading"
      >
        <svg
          height="14"
          viewBox="0 0 10 8"
          width="14"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            fill="currentColor"
            r="1"
            transform="translate(2 4)"
          >
            <animatetransform
              additive="sum"
              attributeName="transform"
              begin="0s"
              dur="1s"
              repeatCount="indefinite"
              type="scale"
              values="0; 1.25; 0"
            />
          </circle>
          <circle
            fill="currentColor"
            r="1"
            transform="translate(5 4)"
          >
            <animatetransform
              additive="sum"
              attributeName="transform"
              begin="0.2s"
              dur="1s"
              repeatCount="indefinite"
              type="scale"
              values="0; 1.25; 0"
            />
          </circle>
          <circle
            fill="currentColor"
            r="1"
            transform="translate(8 4)"
          >
            <animatetransform
              additive="sum"
              attributeName="transform"
              begin="0.4s"
              dur="1s"
              repeatCount="indefinite"
              type="scale"
              values="0; 1.25; 0"
            />
          </circle>
        </svg>
      </span>
      1
    </div>
    <div
      class="stepCount"
      data-status="failed"
    >
      <span
        class="bp3-icon bp3-icon-warning-sign StyledProps--main"
        data-status="failed"
        icon="warning-sign"
      >
        <svg
          data-icon="warning-sign"
          height="14"
          viewBox="0 0 16 16"
          width="14"
        >
          <desc>
            warning-sign
          </desc>
          <path
            d="M15.84 13.5l.01-.01-7-12-.01.01c-.17-.3-.48-.5-.85-.5s-.67.2-.85.5l-.01-.01-7 12 .01.01c-.09.15-.15.31-.15.5 0 .55.45 1 1 1h14c.55 0 1-.45 1-1 0-.19-.06-.35-.15-.5zm-6.85-.51h-2v-2h2v2zm0-3h-2v-5h2v5z"
            fill-rule="evenodd"
          />
        </svg>
      </span>
      1
    </div>
  </div>
</div>
`;
